cluster/partitioner/partitioner.go

```diff
@@ -4,49 +4,64 @@ import (
 	"fmt"
 
 	"github.com/Shopify/sarama"
+	"github.com/cespare/xxhash"
+	jump "github.com/dgryski/go-jump"
 	"github.com/raintank/schema"
 )
 
 type Partitioner interface {
-	Partition(schema.PartitionedMetric, int32) (int32, error)
+	Partition(schema.PartitionedMetric, int32) int32
 }
 
 type Kafka struct {
-	PartitionBy string
-	Partitioner sarama.Partitioner
+	PartitionBy     string
+	Partitioner     sarama.Partitioner
+	GetPartitionKey func(schema.PartitionedMetric, []byte) []byte
 }
 
 func NewKafka(partitionBy string) (*Kafka, error) {
+	kafka := Kafka{
+		PartitionBy: partitionBy,
+	}
+
 	switch partitionBy {
 	case "byOrg":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyByOrgId(b) }
 	case "bySeries":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyBySeries(b) }
+	case "bySeriesWithTags":
+		kafka.Partitioner = sarama.NewManualPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyBySeriesWithTags(b) }
 	default:
-		return nil, fmt.Errorf("partitionBy must be one of 'byOrg|bySeries'. got %s", partitionBy)
+		return nil, fmt.Errorf("partitionBy must be one of 'byOrg|bySeries|bySeriesWithTags'. got %s", partitionBy)
 	}
-	return &Kafka{
-		PartitionBy: partitionBy,
-		Partitioner: sarama.NewHashPartitioner(""),
-	}, nil
+
+	return &kafka, nil
+}
+
+func (k *Kafka) Partition(m schema.PartitionedMetric, numPartitions int32) int32 {
+	key := k.GetPartitionKey(m, nil)
+	return (jumpPartitioner{}).PartitionKey(key, numPartitions)
```

Comment:
it seems like this is the only place where the `jumpPartitioner` is ever used. So what's still the benefit of making it satisfy the `sarama.Partitioner` interface? Due to the fact that it currently implements that interface we need to first instantiate it and then call a method on that instance, that seems unnecessary if it's never used as a sarama partitioner. so we might as well just have a standalone function like `PartitionKeyWithJumpHash([]byte, int32) int32` instead of instantiating this struct and then call the method on it.

Reply:
hmm yes. this and some other things here are rather confusing.
i'll see if i can refactor it.